testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/OIDCClientRegistrationResponseTypesAndGrantsTest.java

```diff
@@ -0,0 +1,243 @@
+        OIDCClientRepresentation clientRep = createRep(List.of(CODE), List.of(AUTHORIZATION_CODE, REFRESH_TOKEN, PASSWORD, CLIENT_CREDENTIALS));
+
+        OIDCClientRepresentation response = reg.oidc().create(clientRep);
+
+        assertOIDCResponse(response, List.of(CODE, NONE), List.of(AUTHORIZATION_CODE, REFRESH_TOKEN, PASSWORD, CLIENT_CREDENTIALS));
+
+        assertKeycloakClient(response, true, false, true, true, true, false);
+    }
+
+    // Grant type "authorization_code" added automatically because of response_type "code" .
+    // If provided response_types are not 100% aligned with provided grant_types, we can in theory reject the request based on https://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata
+    // and the note "The following table lists the correspondence between response_type values that the Client will use and grant_type values that MUST be included in the registered grant_types list"
+    // Not doing it for backwards compatibility for now
+    @Test
+    public void testCodeResponseTypeWithIncompatibleGrants() throws Exception {
+        OIDCClientRepresentation clientRep = createRep(List.of(CODE), List.of(REFRESH_TOKEN, PASSWORD, CLIENT_CREDENTIALS));
+
+        OIDCClientRepresentation response = reg.oidc().create(clientRep);
+
+        assertOIDCResponse(response, List.of(CODE, NONE), List.of(AUTHORIZATION_CODE, REFRESH_TOKEN, PASSWORD, CLIENT_CREDENTIALS));
+
+        assertKeycloakClient(response, true, false, true, true, true, false);
+    }
+
+    private void assertOIDCResponse(OIDCClientRepresentation response, List<String> expectedResponseTypes, List<String> expectedGrantTypes) {
+        assertOIDCResponseImpl(response, expectedResponseTypes, expectedGrantTypes);
+
+        // Test subsequent "get" request returns the same as response from initial "create" request
+        try {
+            reg.auth(Auth.token(response));
+            OIDCClientRepresentation rep = reg.oidc().get(response.getClientId());
+            assertOIDCResponseImpl(rep, expectedResponseTypes, expectedGrantTypes);
+        } catch (ClientRegistrationException cre) {
+            throw new AssertionError(cre);
+        }
+    }
+
+    private void assertOIDCResponseImpl(OIDCClientRepresentation response, List<String> expectedResponseTypes, List<String> expectedGrantTypes) {
+        Assert.assertTrue("Incompatible response_types. Expected response types: " + expectedResponseTypes + ", Received response types: " + response.getResponseTypes(),
+                expectedResponseTypes.size() == response.getResponseTypes().size() && expectedResponseTypes.containsAll(response.getResponseTypes()));
+        Assert.assertTrue("Incompatible grant_Types. Expected grant types: " + expectedGrantTypes + ", Received grant types: " + response.getGrantTypes(),
+                expectedGrantTypes.size() == response.getGrantTypes().size() && expectedGrantTypes.containsAll(response.getGrantTypes()));
```

Comment:
Just a minor but I prefer use the hamcrest matchers:

```java
MatcherAssert.assertThat("Incompatible response_types", response.getResponseTypes(), Matchers.containsInAnyOrder(expectedResponseTypes.toArray()));
MatcherAssert.assertThat("Incompatible grant_Types", response.getGrantTypes(), Matchers.containsInAnyOrder(expectedGrantTypes.toArray()));
```

Reply:
@rmartinc Updated, Thanks!